EVPN Multihoming, EVPN re-factoring for MPLS and L3 Nexthop Group enhancements

AnuradhaKaruppiah opened the pull request:
Based on the discussions in the PR-review meeting we will break this PR up into five separate PRs -
PR-1 (30 commits) - EVPN-MH infra; includes the following sub-features - [mh-prep, EAD, SYNC] - https://github.com/FRRouting/frr/pull/6587
PR-2 (40 commits) - L3NHG infra work
PR-3 (18 commits) - EVPN refactoring for mpls - [refactor-prep, refactor]
PR-4 (~50 commits) - Remaining EVPN MH work - [DF, IRB, feature-fixes]
PR-5 (21 commits) - netns re-work - [netns]

[
This PR is from a topic branch (evpn-mh) used for combining the EVPN-MH, EVPN re-factoring for MPLS and L3 Nexthop group changes.
**EVPN Multihoming** 
[FRR - EVPN Multihoming Overview.pdf](https://github.com/FRRouting/frr/files/4766566/FRR.-.EVPN.Multihoming.Overview.pdf)
This includes support for Active/Active multihoming via the following sub-features -
1. Ethernet Segment configuration and management
2. Type-1/EAD route support
3. Type2 routes with non-zero ESI
4. MAC-ECMP support
5. Layer2 nexthop group support for fast failover of MAC/FDB paths
6. MAC-IP syncing (via type-2 routes)
7. Support for fast failover via local ES-redirect
   PS: And support for slow failover where dataplane doesn't allow for fast failover
8. Host routes (for Symmetric IRB) via Ethernet Segments
9. BGP changes for allowing host routes to be programmed with a L3 nexthop group (again for fast failover)
10. DF election via Type-4/Ethernet Segment routes
11. Support for BUM non-DF and Split-horizon filtering
12. Support for uplink/core tracking
13. Initial support for consistency checking
Signed-off-by: Anuradha Karuppiah <anuradhak@cumulusnetworks.com>

**EVPN re-factoring**
Note this is simply code being around and absolutely NO functional changes were introduced as a part of the re-factoring -
1. Rename of vni to evpn where appropriate
2. move MAC database code from zebra_vxlan.c to zebra_evpn_mac.c
3. move Neighbor database from zebra_vxlan.c to zebra_evpn_neigh.c
4. move core EVPN code from zebra_vxlan.c to zebra_evpn.c
5. Add VXLAN IP learning topotests
Signed-off-by: Pat Ruddy <pat@voltanet.io>

**L3 Nexthop Group**
1. Implement NHG Upper-level Protocol ADD/DEL/REPLACE
2. Divide ID space among all protos and implement ZAPI APIs for ADD/DEL/REPLACEMENT of NHGs in the RIB and dataplane.
3. With this code, sharpd and multihoming BGP can do replacements for Gateway/Ifindex NHGs via 4. one ZAPI and Dataplane message.
5. Upper protos own their ID space and send a full NHG down with the ID space. Zebra tracks them separately in its NHG rib so they are not shared by other protos/zebra.
Signed-off-by: Stephen Worley <sworley@cumulusnetworks.com>

**EVPN NetNS revert and re-work for rebase purposes**
Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
]

louberger commented:
memory usage in LabN BGP CI increases significantly with this PR: 
MASTER: 488 465 426 (this varies about 1-2% per run)
This PR: Memory | 520 523 451

louberger commented:
> memory usage in LabN BGP CI increases significantly with this PR:
> MASTER: 488 465 426 (this varies about 1-2% per run)
> This PR: Memory | 520 523 451

Similar results can bee seen for bgpds in topotest. 
New results: from https://ci1.netdef.org/artifact/FRR-FRRPULLREQ/TP2U1804AMD64/build-12565/TestExecutionLogs/bgp_l3vpn_to_bgp_vrf.test_bgp_l3vpn_to_bgp_vrf/summary.txt

130  ce1    BGPd heap: 7360 KiB --> 41 MiB (0.6728 KiB/vpn route)    1    0
131  ce1    Zebra heap: 8180 KiB --> 33 MiB (0.4964 KiB/vpn route)   1    0
132  ce2    BGPd heap: 7360 KiB --> 41 MiB (0.6728 KiB/vpn route)    1    0
133  ce2    Zebra heap: 8324 KiB --> 37 MiB (0.5735 KiB/vpn route)   1    0
134  ce3    BGPd heap: 7360 KiB --> 36 MiB (0.5728 KiB/vpn route)    1    0
135  ce3    Zebra heap: 7376 KiB --> 31 MiB (0.4725 KiB/vpn route)   1    0
136  r1     BGPd heap: 7640 KiB --> 223 MiB (4.3072 KiB/vpn route)   1    0
137  r1     Zebra heap: 9944 KiB --> 34 MiB (0.4811 KiB/vpn route)   1    0
138  r2     BGPd heap: 7640 KiB --> 205 MiB (3.9472 KiB/vpn route)   1    0
139  r2     Zebra heap: 9980 KiB --> 9980 KiB (0.0 KiB/vpn route)    1    0
140  r3     BGPd heap: 7640 KiB --> 225 MiB (4.3472 KiB/vpn route)   1    0
141  r3     Zebra heap: 9948 KiB --> 35 MiB (0.501 KiB/vpn route)    1    0
142  r4     BGPd heap: 7908 KiB --> 349 MiB (6.8218 KiB/vpn route)   1    0
143  r4     Zebra heap: 9952 KiB --> 58 MiB (0.961 KiB/vpn route)    1    0

Current master: https://ci1.netdef.org/download/FRR-FRR-TOPO2U18AMD64/build_logs/FRR-FRR-TOPO2U18AMD64-3377.log

build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 130  ce1    BGPd heap: 7372 KiB --> 41 MiB (0.6726 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 131  ce1    Zebra heap: 8128 KiB --> 34 MiB (0.5174 KiB/vpn route)   1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 132  ce2    BGPd heap: 7372 KiB --> 41 MiB (0.6726 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 133  ce2    Zebra heap: 7984 KiB --> 65 MiB (1.1403 KiB/vpn route)   1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 134  ce3    BGPd heap: 7372 KiB --> 36 MiB (0.5726 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 135  ce3    Zebra heap: 7324 KiB --> 31 MiB (0.4735 KiB/vpn route)   1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,878 INFO: 136  r1     BGPd heap: 7648 KiB --> 216 MiB (4.167 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 137  r1     Zebra heap: 9892 KiB --> 33 MiB (0.4622 KiB/vpn route)   1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 138  r2     BGPd heap: 7520 KiB --> 197 MiB (3.7896 KiB/vpn route)   1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 139  r2     Zebra heap: 9892 KiB --> 9892 KiB (0.0 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 140  r3     BGPd heap: 7648 KiB --> 218 MiB (4.207 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 141  r3     Zebra heap: 9752 KiB --> 33 MiB (0.465 KiB/vpn route)    1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 142  r4     BGPd heap: 7964 KiB --> 335 MiB (6.5407 KiB/vpn route)   1    0
build	09-Jun-2020 05:41:22	2020-06-09 12:41:22,879 INFO: 143  r4     Zebra heap: 10060 KiB --> 59 MiB (0.9788 KiB/vpn route)  1    0

AnuradhaKaruppiah commented:
> Hi all,
> 
> I talked about this PR last meeting and I think it can be improved to reduce the burden on the authors.
> 
> The reviewability of this PR is terrible for both the developer and the testers:
> 
> Developer side:
> 
> * needs to keep rebasing on new changes
> * keeping track of changes is almost impossible unless you keep bloating the commit count, making rebases/polishing impossible (more commits more possible conflicts)
> * doesn't know if he broke something because he is so far off `master` that it is impossible to tell the old behaviours
> 
> Reviewers side:
> 
> * it is impossible to tell if a commit changed when a rebase happens because of conflicts
> * have to test EVERYTHING EVERTIME time the commits changes
> * it is just not possible to keep all code in mind
> 
> If we go this way it is setting a terrible example for other developers, because:
> 
> 1. it will incentivize bad behavior and we'll be stuck with the following excuse: "oh well, this is too big to review, lets just push in".
> 2. It will also **HIDE** important changes to features other developers made (like zebra data plane that @mjstapp wrote).
>    Theoretical example: how is @mjstapp supposed to review the changes in zebra data plane if it is burried in 180 commits?
> 3. It will cut off community and developers from the development and we'll only be able to review things **AFTER** they got merged.
> 
> I have some experience with this problem because I went through it myself (see the FPM or the northbound PRs: stuck for ~6,7 months with rebase problems).
> 
> > Example: I could have relieved some of my northbound filter changes by backporting some of the CLI fixes... In the end the PR got stuck for 7 months and I had to rebase my code everytime someone fixed things in the old code I had already fixed.
> 
> How can we resolve this? Divide and conquer.
> 
> I went through all commits in this PR and I classified them into separeted classes, and I think we can create separated PRs for these classes so you can:
> 
> 1. Start testing if your changes doesn't break things (even if we are not covering your code)
> 2. Move the burden of resolving conflicts to other developers (e.g. new changes)
> 3. Start to get new code in instead of expecting the all-or-nothing approach.
> 
> ## Commits to Split
> > Reasoning: even though some of these commits are not testable alone, we can test that they do not break a thing. I think it is better to do this way than keep this PR enormous.
> 
> > **TIP:** keep commit header and avoid changing code, so when rebasing this PR git will identify it and not mark as conflict.
> 
> These commits can be commited in another PR to reduce your rebase/maintenance burden:
> 
> (I'll use the commit header instead of SHA because it can change with rebase)
> 
> 1. `lib`/`linux` definitions importation:
> 
> * linux: UAPI for L2-NHG and MAC-ECMP
> * linux: UAPI for MAC sync
> * linux: UAPI for neigh sync
> * lib: bitfield: new macros for bit processing
> * lib: allow listnode memory to be managed by the app
> * lib: timer: move the timer string api to lib
> * lib: Add the ability to grab a nhg starting id
> * zebra: Convert zserv_nexthop_num_warn to return bool
> * lib, bgpd: move json_array_string_add to lib
> * lib: add onlink flag to zapi_nh conversion helper
> * lib,doc: add `onlink` flag to nexthop group config
> * bgpd: sanity check when updating nexthop from bgp to zebra
> 
> 1. Import some data plane mac/vni changes:
> 
> * zebra: support for adding L2NHG and mac-ecmp in the linux kernel
> * zebra: netlink changes for programming MAC-IP SYNC routes in the kernel
> * zebra: support for MAC-IP sync routes
> * zebra: fix misc problems with ES management
> * zebra: rename vni to evpn where appropriate
> * zebra: remove linux/ifbridge.h from rt_netlink.c
> * zebra: Stub rt_socket functions for L2 NHG programming
> 
> 1. Clone and extract required code from other data plane places:
> 
> * zebra: clone `zebra_vxlan.c` to `zebra_evpn_mac.c` (spurious comment in `zebra/zebra_vxlan.c`)
> * zebra: extract evpn mac functions from zebra_vxlan.c
> * zebra: split out `mac_add` code from `process_remote_macip_add`
> * zebra: extract local mac add code from vxlan
> * zebra: extract local mac del from `zebra_vxlan.c`
> * zebra: clone `zebra_vxlan.c` to `zebra_evpn_neigh.c` (spurious comment in `zebra/zebra_vxlan.c`)
> * zebra: extract neighbor functions from zebra_vxlan.c
> * zebra: extract neigbor processing from `remote_macip_add`
> * zebra: extract neigbor processing from `zevpn_gw_macip_add`
> * zebra: extract neighbor processing from `process_remote_macip_del`
> * zebra: extract neighbor processing from `kernel_neigh_del`
> * zebra: clone `zebra_vxlan.c` to `zebra_evpn.c` (spurious comment in `zebra/zebra_vxlan.c`)
> * zebra: extract core EVPN functions from `zebra_vxlan.c`
> * zebra: display interface slave type
> * zebra: remove a stray comment (this commit remove some previous spurios comments, so squash this)
> * zebra: fixup problem with neigh "sync" flag setting
> 
> 1. Zebra namespace handling code:
> 
> * zebra: importation of bgp evpn rt5 from vni with other netns
> * zebra: `zvni_from_svi()` adaptation for other network namespaces
> * zebra: `zvni_map_to_svi()` adaptation for other network namespaces
> * zebra: map vxlan interface to bridge interface with correct ns id
> * zebra, lib: add an internal API to get relative default nsid in other ns
> * zebra, lib: store relative default ns id in each namespace
> * zebra, lib: new API to get absolute netns val from relative netns val
> * zebra: fdb and neighbor table are read for all zns
> * zebra: bridge layer2 information records ns_id where bridge is
> * zebra: add ns_id attribute to mac structure
> * zebra: zvni_map_to_vlan() adaptation for all namespaces
> * zebra: local mac entries populated in correct netnamespace
> * lib, zebra: reuse and adapt ns_list walk functionality
> * zebra: dynamically detect vxlan link interfaces in other netns
> * doc: add some documentation about bgp evpn netns support
> 
> 1. More EVPN code importation into zebra
> 
> * zebra: changes for configuring mac and neigh holdtime
> * zebra: setup static neigh flag in the dataplane
> * zebra: remove FDB entries before de-activating a L2-NHG
> * zebra: changes to run DF election
> * zebra: changes for programming SPH, non-DF and backup NHG br-port attrs
> * zebra: handle local-es bridge port association
> * zebra: dplane APIs for programming evpn-mh access port attributes
> * zebra: fixup the previous commit (squash this commit into the previous)
> * zebra: support for slow-failover of local MACs on an ES
> * zebra: uplink tracking and startup delay for EVPN-MH
> * zebra: uplink tracking fixups
> * zebra: fix fdb netlink update handler
> * zebra: stop neigh hold timer when the neigh is deleted
> * zebra: debug logs to detect incorrect mac deletions
> 
> 1. Zebra Next Hop Group related code.
> 
> * zebra: Refactor nexthop reading from zapi messages
> * lib, zebra: Add ZAPI_NHG_ADD|DELETE
> * sharpd: add abilty to send a nhg to zebra
> * zebra: remove NHG unhashable flag and its code
> * lib, zebra: Add ability to send down a nhgid over route install
> * zebra: implement protocol NHG Add/Del
> * zebra: free up the L2 NHG bitmap as a part of shutdown
> * zebra: Set NUD_NOARP on sticky MAC entries in addition to NTF_STICKY
> * zebra: fixup to uplink tracking
> * zebra: hash proto-created but zebra ID spaced NHGS
> * zebra: implement NHG proto replace
> * zebra: handle zapi routes with NHG ID set
> * zebra: special handling for proto-NHG-based routes
> * zebra: dont update counter if outside of zebra ID range
> * zebra: NHG ID bounds macros
> * sharpd: print the correct ID the NHG is using
> * zebra,lib: command to only install proto-based nexthops
> * zebra: make NHG ID allocation smarter
> * zebra: use the passed proto from zapi
> * zebra: fix route validity check with NHG ID
> * zebra: change the L2 NHG id format to co-exist with the L3NHG ids
> * zebra: Prevent duplicate re-install
> * zebra: skip NDA_DST attr if NHG is present
> * zebra: change the nhg format from hex to dec for easy match up with t
> * zebra: remove unneeded nhg repalce boilerplate (this commit may be squashed)
> * zebra: inc/dec refcount on add/del NHG proto
> * zebra: handle proto NHG uninstall client disconnect
> * zebra: warn if zapi NHG add has no nexthops
> * zebra: fix refcnt/rib issues in NHG replace/delete
> * zebra: fixup pending items
> * zebra: increment the nhg proto score iterator
> * zebra: fix releasing proto-owned singletons
> * zebra: return the proto nhe on del even with refs
> * lib,sharpd,pbrd: `set installable` nhg command
> * lib: add logging for ZEBRA_NHG_ADD[DEL]
> * zebra: add logging for NHG ignoring in netlink
> * sharpd: replace strncpy with strlcpy (or squash it)
> * zebra: Only install a minimal amount of times
> * zebra: Properly set NEXTHOP_FLAG_FIB when skipping install
> 
> 1. Small BGP changes to accomodade incomping EVPN multihoming.
> 
> * bgpd: pull the multihoming code out to a separate file
> * bgpd: fix ecom leak handling l3vni update
> 
> ## Commits Change
> I think all commits with `Revert FOO` should be squashed into a single one and you should comment why they were reverted.
> 
> Specific commits:
> 
> * bgpd: Free up the NHG id resources last (this commit may be squashed)
> * bgpd: fixup CLANG warnings (this commit may be squashed or edited in rebase)
> * zebra: fixup clang warnings (this commit should be squashed/rebased)
> * zebra: fixup a neigh del bug introduced by a previous commit for MAC (this commit should be squashed/rebased)
> * bgpd: changes for DF election (there are `lib` changes that can be commited separatedly)
> * zebra: re-name some mh functions to make the code more readable (may be squashed)
> * zebra: fix deletion of evpn mh neigh-holdtime (may be squashed)
> * Revert "lib,sharpd,pbrd: `set installable` nhg command"  (should be squashed)
> * bgp: remove duplicate command installs  (maybe squashed or moved to another PR)
> * remove strcpy, strlen and sprintf calls (maybe squashed or rebase to apply to specific commits)

@rzalamena, Thanks for taking the time to go over the PR and providing your feedback. I understand the angst behind the comments. And I am certainly going to try and address them the best I can -
1. Of all the features I have implemented I have to say EVPN-Multihoming is the most challenging in terms of both the complexity of the feature and the sheer amount of work needed to make it a deploy-able solution. I have no doubt reviewing it will be a massive challenge as well. To that end I have attached a detailed functional specification to the PR. 
2.  Re-ordering commits and sending them as multiple PRs is not practical. The flow of the code will be lost. Instead I am going to group the commits (along the same lines as you suggest) and tag relevant groups with corresponding experts.
3. Most of the initial fixes were already squashed pre-upstreaming (infact hundreds of commits were already squashed) but after the refactoring it became impractical. I will audit the commits you pointed at and work my co-authors to have them squashed/fixed-up to the extent possible.

AnuradhaKaruppiah commented:
> > memory usage in LabN BGP CI increases significantly with this PR:
> > MASTER: 488 465 426 (this varies about 1-2% per run)
> > This PR: Memory | 520 523 451
> 
> Similar results can bee seen for bgpds in topotest.
> New results: from https://ci1.netdef.org/artifact/FRR-FRRPULLREQ/TP2U1804AMD64/build-12565/TestExecutionLogs/bgp_l3vpn_to_bgp_vrf.test_bgp_l3vpn_to_bgp_vrf/summary.txt
> 
> 130 ce1 BGPd heap: 7360 KiB --> 41 MiB (0.6728 KiB/vpn route) 1 0
> 131 ce1 Zebra heap: 8180 KiB --> 33 MiB (0.4964 KiB/vpn route) 1 0
> 132 ce2 BGPd heap: 7360 KiB --> 41 MiB (0.6728 KiB/vpn route) 1 0
> 133 ce2 Zebra heap: 8324 KiB --> 37 MiB (0.5735 KiB/vpn route) 1 0
> 134 ce3 BGPd heap: 7360 KiB --> 36 MiB (0.5728 KiB/vpn route) 1 0
> 135 ce3 Zebra heap: 7376 KiB --> 31 MiB (0.4725 KiB/vpn route) 1 0
> 136 r1 BGPd heap: 7640 KiB --> 223 MiB (4.3072 KiB/vpn route) 1 0
> 137 r1 Zebra heap: 9944 KiB --> 34 MiB (0.4811 KiB/vpn route) 1 0
> 138 r2 BGPd heap: 7640 KiB --> 205 MiB (3.9472 KiB/vpn route) 1 0
> 139 r2 Zebra heap: 9980 KiB --> 9980 KiB (0.0 KiB/vpn route) 1 0
> 140 r3 BGPd heap: 7640 KiB --> 225 MiB (4.3472 KiB/vpn route) 1 0
> 141 r3 Zebra heap: 9948 KiB --> 35 MiB (0.501 KiB/vpn route) 1 0
> 142 r4 BGPd heap: 7908 KiB --> 349 MiB (6.8218 KiB/vpn route) 1 0
> 143 r4 Zebra heap: 9952 KiB --> 58 MiB (0.961 KiB/vpn route) 1 0
> 
> Current master: https://ci1.netdef.org/download/FRR-FRR-TOPO2U18AMD64/build_logs/FRR-FRR-TOPO2U18AMD64-3377.log
> 
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 130 ce1 BGPd heap: 7372 KiB --> 41 MiB (0.6726 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 131 ce1 Zebra heap: 8128 KiB --> 34 MiB (0.5174 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 132 ce2 BGPd heap: 7372 KiB --> 41 MiB (0.6726 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 133 ce2 Zebra heap: 7984 KiB --> 65 MiB (1.1403 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 134 ce3 BGPd heap: 7372 KiB --> 36 MiB (0.5726 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 135 ce3 Zebra heap: 7324 KiB --> 31 MiB (0.4735 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,878 INFO: 136 r1 BGPd heap: 7648 KiB --> 216 MiB (4.167 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 137 r1 Zebra heap: 9892 KiB --> 33 MiB (0.4622 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 138 r2 BGPd heap: 7520 KiB --> 197 MiB (3.7896 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 139 r2 Zebra heap: 9892 KiB --> 9892 KiB (0.0 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 140 r3 BGPd heap: 7648 KiB --> 218 MiB (4.207 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 141 r3 Zebra heap: 9752 KiB --> 33 MiB (0.465 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 142 r4 BGPd heap: 7964 KiB --> 335 MiB (6.5407 KiB/vpn route) 1 0
> build 09-Jun-2020 05:41:22 2020-06-09 12:41:22,879 INFO: 143 r4 Zebra heap: 10060 KiB --> 59 MiB (0.9788 KiB/vpn route) 1 0

@louberger I will work with @donaldsharp  and @sworleys  on this and we will absolutely get to the bottom of it.

rzalamena commented:
@AnuradhaKaruppiah 

> @rzalamena, Thanks for taking the time to go over the PR and providing your feedback. I understand the angst behind the comments. And I am certainly going to try and address them the best I can -
> 1. Of all the features I have implemented I have to say EVPN-Multihoming is the most challenging in terms of both the complexity of the feature and the sheer amount of work needed to make it a deploy-able solution. I have no doubt reviewing it will be a massive challenge as well. To that end I have attached a detailed functional specification to the PR. 

I'm not questioning the technical complexity of the feature, I'm warning whoever is going to maintain the branch that if some of these commits are not cherry picked into a separated PR there will be a lot of work hours wasted for the reasons I explained.

It will also make the life of reviewers painful (or easy if we go the bad route) because the amount of code/commits.


> 2.  Re-ordering commits and sending them as multiple PRs is not practical. The flow of the code will be lost. Instead I am going to group the commits (along the same lines as you suggest) and tag relevant groups with corresponding experts.

I just reordered the commits so they can be commited in separated PRs. Do you want me to start with the first class (`lib/linux definitions importation`) to set as an example?

It will remove 12 commits from the PR and will allow other people to start reviewing those trivial changes. If no code changes it will also be very trivial to rebase, git will just skip the commits from the new `master`.

The commits (most of them) are in great shape. They are mostly contained (e.g. don't edit multiple files at the same time) and that is why I think it will be less work to get some of them first before this PR.

Like I said: we can get some commits in even though some functions are still not in use (I'd just want to test it doesn't break anything or changes unintended behavior, as a plus it will prevent other developers from doing conflicting changes).


> 3. Most of the initial fixes were already squashed pre-upstreaming (infact hundreds of commits were already squashed) but after the refactoring it became impractical. I will audit the commits you pointed at and work my co-authors to have them squashed/fixed-up to the extent possible.

Some of the commits I've pointed out might not be trivial (because I might have missed dependencies or order), however there are others that are easily doable (see example below).

Squashing the `revert` commits is also trivial, after the squash happens git will open an editor for you to edit the new commit message and the author can tell us why those have been reverted.

> ![revert-commits](https://user-images.githubusercontent.com/205063/84510968-5eeefc80-acb5-11ea-8488-385d06b213c8.png)

Because messages like `This reverts commit FOO.` are not helpful in this context and if there is a conflict in one of those `reverts` you'll have to edit all other commits after that.

Please let me know if someone needs help doing some `git rebase`s, I might have some tricks to help (re)organize the commits.

AnuradhaKaruppiah commented:
> @AnuradhaKaruppiah
> 
> > @rzalamena, Thanks for taking the time to go over the PR and providing your feedback. I understand the angst behind the comments. And I am certainly going to try and address them the best I can -
> > 
> > 1. Of all the features I have implemented I have to say EVPN-Multihoming is the most challenging in terms of both the complexity of the feature and the sheer amount of work needed to make it a deploy-able solution. I have no doubt reviewing it will be a massive challenge as well. To that end I have attached a detailed functional specification to the PR.
> 
> I'm not questioning the technical complexity of the feature, I'm warning whoever is going to maintain the branch that if some of these commits are not cherry picked into a separated PR there will be a lot of work hours wasted for the reasons I explained.
> 
> It will also make the life of reviewers painful (or easy if we go the bad route) because the amount of code/commits.
> 
> > 1. Re-ordering commits and sending them as multiple PRs is not practical. The flow of the code will be lost. Instead I am going to group the commits (along the same lines as you suggest) and tag relevant groups with corresponding experts.
> 
> I just reordered the commits so they can be commited in separated PRs. Do you want me to start with the first class (`lib/linux definitions importation`) to set as an example?
> 
> It will remove 12 commits from the PR and will allow other people to start reviewing those trivial changes. If no code changes it will also be very trivial to rebase, git will just skip the commits from the new `master`.
> 
> The commits (most of them) are in great shape. They are mostly contained (e.g. don't edit multiple files at the same time) and that is why I think it will be less work to get some of them first before this PR.
> 
> Like I said: we can get some commits in even though some functions are still not in use (I'd just want to test it doesn't break anything or changes unintended behavior, as a plus it will prevent other developers from doing conflicting changes).
> 
> > 1. Most of the initial fixes were already squashed pre-upstreaming (infact hundreds of commits were already squashed) but after the refactoring it became impractical. I will audit the commits you pointed at and work my co-authors to have them squashed/fixed-up to the extent possible.
> 
> Some of the commits I've pointed out might not be trivial (because I might have missed dependencies or order), however there are others that are easily doable (see example below).
> 
> Squashing the `revert` commits is also trivial, after the squash happens git will open an editor for you to edit the new commit message and the author can tell us why those have been reverted.
> 
> > ![revert-commits](https://user-images.githubusercontent.com/205063/84510968-5eeefc80-acb5-11ea-8488-385d06b213c8.png)
> 
> Because messages like `This reverts commit FOO.` are not helpful in this context and if there is a conflict in one of those `reverts` you'll have to edit all other commits after that.
> 
> Please let me know if someone needs help doing some `git rebase`s, I might have some tricks to help (re)organize the commits.
@rzalamena  I will reflect on the PR breakup. And I am already working with my co-authors on the squashing, stay tuned.

donaldsharp commented:
@louberger I see two data structure size differences
`struct bgp_node` is now 200 bytes -vs- 192
This is because `struct route_node` has grown by 8 bytes in the `union u` because `struct evpn_addr` has grown by 8 bytes( which is the biggest part of the union ).  This is because `struct evpn_ead_addr` has grown by 8 bytes.

`struct bgp_path_info_extra` has grown by 8 bytes because of the addition of the`struct bgp_path_es_info *es_info` pointer.

I believe this is why bgp size has grown.

donaldsharp commented:
With the latest commit,the bgp_l3vpn_to_bgp_vrf test memory increase goes from 340mb -> 346 mb @louberger can you look at your CI systems to see if they are better now?

AnuradhaKaruppiah commented:
Based on the discussion we will break this PR up into five separate PRs -
1. PR-1 (30 commits)  - EVPN-MH infra; includes the following sub-features - [mh-prep, EAD, SYNC]
PS: netns changes will be reverted as a part of [mh-prep]
2. PR-2 (40 commits) - L3NHG infra work
3. PR-3 (18 commits) - EVPN refactoring for mpls - [refactor-prep, refactor]
4. PR-4 (~50 commits) - Remaining EVPN MH work - [DF, IRB, feature-fixes]
5. PR-5 (21 commits) - netns re-work - [netns] 

I will send out PR-1 shortly and link it to this PR (for anybody who wants to see what else is in the pipeline).

Spantik commented:
replaced by #6587
